### DISCUSSION TITLE
Smaller size? (adds 66kb)

### DISCUSSION THREAD

**Discussion by mustafa0x** [focus-trap/focus-trap#1254] [General]
I ran https://github.com/btd/rollup-plugin-visualizer on my bundle, and I noticed that focus-trap (40kb) + tabbable (20kb) take up 66kb (before minifiying). Which is a quite a bit, all things considered.

Is is possible that this utility will be made more lean in the future?

Thanks for the useful tool.

**Comment by stefcameron** [comment]
@mustafa0x I'm glad you find the libraries useful! A smaller pre-minification footprint is not something I plan to focus on any time soon, though I realize the code is (very) old and could stand to be optimized.

But as you say, this is before minification, and I would expect you're using the minified code in Prod. I'm sure it comes down quite a bit after minification since the code currently uses some really long variable names, for example. And the pre-minified code could be further optimized (in size reduction) by replacing some older, less succinct code patterns that have since been simplified.

**Comment by stijns96** [comment]
Hi @stefcameron,

Any plans to make it more modern? I also found out that the code is really large for something like a trap focus.

**Reply by stefcameron** [reply]
Can you define "really large" and confirm whether this is before or after minification and gzipping?

I certainly don't discount the possibility of refactoring the code being of some help for post-minification size, but I think it would require a huge amount of time/effort for what would be quite small gains in the end.

Some other thoughts to consider:

- This library is not targeting/supporting any mobile platforms. Desktop only (with presumably much faster connections and processing, in contrast to mobile).
- If you're comparing this library to modern browser-native features like the [inert](https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/inert) attribute, that will always win. This library was conceived years before `inert` existed and sought to address a gap. Turns out there were a ton of use cases around trapping focus, many of which `inert` by itself doesn't address.
- Completely refactoring the implementation so it's inert-based isn't possible because of existing options like `allowOutsideClick` or `clickOutsideDeactivates`. e.g. How would use `inert` to render the page inert while floating an unaffected `div` providing a dialog UI above it all, and somehow still receive clicks on some inert node outside that `div`...?
- The size of this library is almost in direct proportion to the number of [options](https://github.com/focus-trap/focus-trap?tab=readme-ov-file#createoptions) (24 by my count!) it has grown to support. Perhaps the library failed to remain on a very narrow path of options to support and, over years, has grown to be bloated because of it, but dropping any one of them would affect _someone_ out there trying to use Focus-trap. You'd be amazed at how many different use cases there are out there for/when trapping focus.
- A few of these options support a bunch of different values, e.g. `initialFocus` with `HTMLElement | SVGElement | string | false | undefined | (() => HTMLElement | SVGElement | string | false | undefined)` -- handling each of these adds a lot of flexibility, but also a lot more code.

I do have _hopes_ (not necessarily _plans_) to convert the library to native TypeScript, but I don't see that resulting in a significant pre-minification (even post, for that matter) size reduction.

**Reply by stijns96** [reply]
Hi @stefcameron 

Thanks for the very detailed explanation! 

We currently have our own trap focus util that is missing some core features. After implemention this module, I was somehow quite shocked about the size of it.

We use Rollup with treeshaking that didn't seem to do anything. Also, we use this for shopify themes, so gzip and minification is done on their end.

Not pretty sure about the typescript part, but I got hooked on the fact that you actually admid it could be more modern.

I'll take another look later on to check the treeshaking etc.

Again, thanks for taking the time answering this :)

**Reply by stefcameron** [reply]
@stijns96 You're welcome! 😄 

> We use Rollup with treeshaking that didn't seem to do anything.

If you do additional tests and you find places in the code where some treeshaking pragma comments could be added to help bundlers to a better job with the Focus-trap code, I'd be happy to take a PR to add those in!

**Reply by mustafa0x** [reply]
> This library is not targeting/supporting any mobile platforms. Desktop only (with presumably much faster connections and processing, in contrast to mobile).

Even so, due to bundling, it'll always be included in 99% of cases. :)

And it's often a sub-dep of the ui lib the dev is using. So it can't easily be removed even if the dev wanted to.

**Comment by Reeska** [comment]
Hi,
If I may, you can check the final minified gzipped bundle size here : https://bundlephobia.com/package/focus-trap@7.6.3
It's about 5.8 kB, it seems fine to me.

**Reply by mustafa0x** [reply]
Missing tabbable, which is 2.2kb.

So ~22kb minified, and 8kb gzipped. It's quite a lot to add to every single web app on the web. Especially since the core features can be written in <1kb.

I'm sure more is needed to catch all cases. My question is, can we find a reasonable middle ground?
